### PR TITLE
**Feature:** GO HARD on a11y

### PR DESCRIPTION
**Pull request by TejasQ**
# Summary
This PR fixes a number of issues with Inputs:
- Odd behavior when used without/with labels
- Fix spacing when an Input has an error AND a label
- Fix cursors on Input buttons
- Refactor legacy `styled` import
- Tooltips now have `id`s for screen readers
- Make all `Input`s accessible: with `aria` properties and correct label handling
- Update all examples
- Refactor `Autocomplete` to use hooks and support IDs

# To Do
- [x] Make the same changes to `Textarea`

# Open separate PRs for
- [ ] Make the same changes to `Select`
- [ ] Make the same changes to `Checkbox`
- [ ] Make the same changes to `Switch`
- [ ] Make the same changes to `Toggle`

# To be tested

Me
- [x] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/Autocomplete/Autocomplete.tsx [hunk 1]
```diff
@@ -70,52 +76,42 @@ const Container = styled(ContextMenu)<{ fullWidth: boolean }>`
...
+  resultIcon,
+  loading,
+  noResultsMessage = "No Results Found",
+  onResultClick,
+  selectedResult,
+  children,
+  ...inputProps
+}: AutocompleteProps<T>) {
+  const [isContextMenuOpen, setIsContextMenuOpen] = React.useState(false)
+  const autocompleteId = React.useRef(id || nanoid())
```
Comment (stereobooster) [inline src/Autocomplete/Autocomplete.tsx]:
```ts
const useUniqueId = () => {
  const id = React.useRef<string>(null);
  if (!id.current) id.current = nanoid();
  return id.current;
}
```

Comment (TejasQ) [reply src/Autocomplete/Autocomplete.tsx]: UFUFUFFUFUFUFU

File: src/Tooltip/Tooltip.tsx [hunk 2]
```diff
@@ -179,13 +180,14 @@ const Tooltip: React.SFC<TooltipProps> = props => {
         <p>{props.children}</p>
       </Container>
       <Container
+        aria-role="tooltip"
```
Comment (stereobooster) [inline src/Tooltip/Tooltip.tsx]:
```suggestion
        role="tooltip"
```

File: src/useUniqueId/index.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import nanoid from "nanoid"
+import { useRef } from "react"
+
+export const useUniqueId = (defaultId: string | null = null) => {
```
Comment (stereobooster) [inline src/useUniqueId/index.ts]:
```ts
export const useUniqueId = (defaultId: string | null = null) => React.useState(nanoid)[0] as string;
```

source https://dev.to/alutom/comment/984j

**Comment by stereobooster** [comment]
Does snapshot tests fail because of this https://github.com/facebook/jest/issues/7094?

File: src/Input/Input.Button.tsx [hunk 4]
```diff
@@ -35,6 +35,9 @@ const Button = styled("div")`
   align-items: center;
   justify-content: center;
   cursor: pointer;
+  * {
```
Comment (stereobooster) [inline src/Input/Input.Button.tsx]: why?

File: src/Input/Input.Field.tsx [hunk 5]
```diff
@@ -128,39 +127,41 @@ const InputField: React.SFC<InputProps> = ({
...
-        {clear && value && (
-          <ClearButton onClick={clear}>
-            <Icon color="color.text.lightest" name="No" />
-          </ClearButton>
-        )}
-      </Container>
+    <Container fullWidth={fullWidth} withLabel={Boolean(label)}>
+      {shouldShowIconButton && renderButton()}
+      <Field
+        aria-labelledby={label ? `input-label-${id}` : undefined}
```
Comment (stereobooster) [inline src/Input/Input.Field.tsx]: `ariaLabelledby`?

File: src/utils/mixins.ts [hunk 6]
```diff
@@ -55,12 +55,15 @@ export const FormFieldControl = styled("div")(({ theme }) => ({
   ":hover": {
     color: theme.color.text.default,
   },
+  " *": {
```
Comment (stereobooster) [inline src/utils/mixins.ts]: Is it the same as `"& *": {`?

File: src/Tooltip/Tooltip.tsx [hunk 7]
```diff
@@ -16,6 +16,7 @@ export interface BaseProps extends DefaultProps {
   children?: React.ReactNode
   /** Smart-positioned tooltip, with positioning reversed so it doesn't flow out of the window's bounding box. Currently works for left and top-positioned tooltips. */
   smart?: boolean
+  textId?: string
```
Comment (stereobooster) [inline src/Tooltip/Tooltip.tsx]:
```suggestion
id?: string
```
?

File: src/Textarea/Textarea.tsx [hunk 8]
```diff
@@ -1,18 +1,21 @@
...
+import { useUniqueId } from "../useUniqueId"
+import { isCmdEnter, lighten } from "../utils"
 import { FormFieldControls, FormFieldError, inputFocus, Label } from "../utils/mixins"
+import styled from "../utils/styled"
 
 type ResizeOptions = "none" | "both" | "vertical" | "horizontal"
 
 export interface TextareaProps extends DefaultProps {
+  /** What is the identifier of this textarea? */
+  id?: string
```
Comment (stereobooster) [inline src/Textarea/Textarea.tsx]: `tabIndex`? Maybe introduce `type AccessibleInput = { ... }` and then reuse it across input, textarea, autocomplete 🤔

**Comment by stereobooster** [comment]
lock icon is not focusable
Screenshot 2019-03-11 at 15 41 13

I'm not sure how this suppose to work. Copy button is not focusable.
Screenshot 2019-03-11 at 15 41 05

File: src/Input/Input.Field.tsx [hunk 9]
```diff
@@ -128,39 +127,41 @@ const InputField: React.SFC<InputProps> = ({
...
+        aria-describedby={hint ? `input-hint-${id}` : undefined}
+        aria-label={label}
```
Comment (stereobooster) [inline src/Input/Input.Field.tsx]: do we need both `aria-labelledby` and `aria-label`?

File: src/Input/Input.Field.tsx (hunk 5 above)
Comment (stereobooster) [reply src/Input/Input.Field.tsx]: Also move it to `Input/Input.tsx` so 

```tsx
aria-describedby={hint ? `input-hint-${id}` : undefined}
```

would be in the same file as 

```tsx
{hint && <Hint textId={`input-hint-${uniqueId}`}>{hint}</Hint>}
```

File: src/utils/mixins.ts (hunk 6 above)
Comment (TejasQ) [reply src/utils/mixins.ts]: No. It is equivalent to `"> *"` instead. 🙂

File: src/Tooltip/Tooltip.tsx (hunk 7 above)
Comment (TejasQ) [reply src/Tooltip/Tooltip.tsx]: Naaaaa

**Pull request merged**